### PR TITLE
Cirrus: Fix gate image & false-positive exits

### PR DESCRIPTION
**Pull request by cevich**
***Depends on:*** #5066 

Signed-off-by: Chris Evich <cevich@redhat.com>

### CODE REVIEW

**Comment by mheon** [comment]
/approve
LGTM but would like someone more familiar with bash to review

File: Makefile [hunk 1]
```diff
@@ -104,6 +104,8 @@ GOBIN := $(shell $(GO) env GOBIN)
 ifeq ($(GOBIN),)
 GOBIN := $(FIRST_GOPATH)/bin
 endif
+# Required by some scripts
+export GOBIN
```
Comment (TomSweeneyRedHat) [inline Makefile]: can you do this export  if it's not defined?  CI tests aren't happy.

Comment (cevich) [reply Makefile]: removed

**Comment by cevich** [comment]
> LGTM but would like someone more familiar with bash to review

Yeah, that and I need to improve the commit message and add some comments...

**Comment by cevich** [comment]
@mheon @baude FYI- I still have some i-dotting and t-crossing to do here.  For the moment, it looks like whatever was causing download failures has passed, so I think there is less urgency to this PR.  However, I did uncover and fix a bunch of false-positive problems in our current setup.

- [ ] Confirm the `success` job works with the new gate container
- [ ] Confirm w/n a manual image build & push is needed or if the current master and PR usage of the new image will work as-is.
- [ ] Notify people to rebase their PRs if required once the new image is built and pushed.

**Comment by cevich** [comment]
Note: There are a few documentation issues I need to clean up and additional local testing needed (see checklist above)

**Comment by cevich** [comment]
@mheon @baude @edsantiago manual testing has shown, we will get gate-jobs failing for a short duration once this PR merges.  However, once quay builds the new image, re-running any failed jobs should fix them.

FWIW: I did experiment with trying the new code using the old container, and visa-versa but couldn't find a sane way to keep all use-cases happy.

File: .cirrus.yml [hunk 2]
```diff
@@ -85,17 +86,18 @@ gce_instance:
...
 
     env:
         CIRRUS_WORKING_DIR: "/usr/src/libpod"
-        GOPATH: "/go"
-        GOSRC: "/go/src/github.com/containers/libpod"
+        SRCPATH: "$CIRRUS_WORKING_DIR"
 
     # Runs within Cirrus's "community cluster"
     container:
+        # Note: Image has dual-purpose, see contrib/gate/README.md
```
Comment (edsantiago) [inline .cirrus.yml]: Nit: improper hyphen (should just be "Image has dual purpose"). Two other instances below.

File: contrib/gate/Dockerfile [hunk 3]
```diff
@@ -33,31 +33,36 @@ RUN dnf -y install \
...
+# Only needed for installing build-time dependencies, then will be removed
 COPY / $GOSRC
 
-WORKDIR $GOSRC
-
 # Install dependencies
 RUN set -x && \
+    mkdir -p "$GOBIN" && \
+    mkdir -p /etc/cni/net.d && \
+    mkdir -p /etc/containers && \
```
Comment (edsantiago) [inline contrib/gate/Dockerfile]: Why this change? I find it more maintainable to group each mkdir with its install.

File: contrib/gate/entrypoint.sh [hunk 4]
```diff
@@ -1,15 +1,23 @@
 #!/bin/bash
 
-[[ -n "$SRCPATH" ]] || \
-    ( echo "ERROR: \$SRCPATH must be non-empty" && exit 1 )
-[[ -n "$GOSRC" ]] || \
-    ( echo "ERROR: \$GOSRC must be non-empty" && exit 2 )
+set -e
+
+die() {
+    echo "${2:-FATAL ERROR (but no message given!) in ${FUNCNAME[1]}()}"
```
Comment (edsantiago) [inline contrib/gate/entrypoint.sh]: A nice habit to develop is embedding a helpful prefix in `die` messages: something that explains where the message is coming from. Usually `$(basename $0)` is good, but in this case `entrypoint.sh` is useless so it should be something like `gate` or `gate-container` or something that will help a poor user track down where the error is actually coming from.

File: contrib/gate/entrypoint.sh [hunk 5]
```diff
@@ -1,15 +1,23 @@
...
+    exit ${1:-1}
+}
+
+[[ -n "$SRCPATH" ]] || die 1 "ERROR: \$SRCPATH must be non-empty"
+[[ -n "$GOPATH" ]] || die 2 "ERROR: \$$GOPATH must be non-empty"
```
Comment (edsantiago) [inline contrib/gate/entrypoint.sh]: Too many dollars.

File: hack/get_release_info.sh [hunk 6]
```diff
@@ -7,7 +7,7 @@
 set -euo pipefail
 
 DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
```
Comment (edsantiago) [inline hack/get_release_info.sh]: DIR is no longer used, so please get rid of it

File: hack/install_golangci.sh [hunk 7]
```diff
@@ -1,17 +1,16 @@
...
-	echo \$VERSION is empty
-	exit 1
-fi
+set -e
 
-if [ -z "$GOBIN" ]; then
-	echo \$GOBIN is empty
-	exit 1
-fi
+die() { echo "${1:-No error message given}"; exit 1; }
```
Comment (edsantiago) [inline hack/install_golangci.sh]: This is a good place to use `$(basename $0)` as a prefix in the `die` message.

File: hack/install_golangci.sh [hunk 8]
```diff
@@ -1,17 +1,16 @@
...
-$GOBIN/golangci-lint --version | grep $VERSION
-if [ $?  -ne 0 ]; then
-	set -e
-	curl -sfL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s -- -b $GOBIN v$VERSION
+BIN="$GOBIN/golangci-lint"
+if [ ! -x "$BIN" ]; then
+    echo "Installing golangci-lint v$VERSION into $GOBIN"
+    curl -sfL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s -- -b $GOBIN v$VERSION
+else
+    echo "Using existing golangci-lint v$($BIN --version | grep "$VERSION") at $BIN"
```
Comment (edsantiago) [inline hack/install_golangci.sh]: Whoa! What is this supposed to do? My guess is you want something like:
```bash
   echo "$BIN already installed: $($BIN --version)"
```

File: Makefile [hunk 9]
```diff
@@ -536,7 +536,7 @@ uninstall:
 .PHONY: .gitvalidation
 .gitvalidation: .gopathok
 	@echo "Validating vs commit '$(call err_if_empty,EPOCH_TEST_COMMIT)'"
-	GIT_CHECK_EXCLUDE="./vendor:docs/make.bat" $(GOBIN)/git-validation -run DCO,short-subject,dangling-whitespace -range $(EPOCH_TEST_COMMIT)..$(HEAD)
+	GIT_CHECK_EXCLUDE="./vendor:docs/make.bat" $(GOBIN)/git-validation -v -run DCO,short-subject,dangling-whitespace -range $(EPOCH_TEST_COMMIT)..$(HEAD)
```
[the remark this answers is not in the document]
Comment (cevich) [reply Makefile]: oh right, good catch.  I thought that would help make the 'bad commit' EPOCH_TEST_COMMIT error more informative. But I just tried again, and even with `-v` it still only exits 128 w/o a messsage :disappointed:

File: .cirrus.yml (hunk 2 above)
Comment (cevich) [reply .cirrus.yml]: got it.

File: contrib/gate/Dockerfile (hunk 3 above)
Comment (cevich) [reply contrib/gate/Dockerfile]: There are fewer layers to download this way.

File: contrib/gate/entrypoint.sh (hunk 4 above)
Comment (cevich) [reply contrib/gate/entrypoint.sh]: Great suggestion.

File: hack/install_golangci.sh (hunk 8 above)
Comment (cevich) [reply hack/install_golangci.sh]: wrong copy-paste from earlier version.  Fixing...

**Comment by edsantiago** [comment]
Looks like a real failure in `success.sh`:
```
contrib/cirrus/lib.sh: line 41: cd: /var/tmp/go/src/github.com/containers/libpod: No such file or directory
```
There are too many containers and contexts for me to figure out the right solution for this one; I'll trust that it's easy for you to do so.

**Comment by cevich** [comment]
> There are too many containers and contexts for me to figure out the right solution for this one; I'll trust that it's easy for you to do so.

This is simply a reflection of what I confirmed earlier.  Due to unification of paths used in the gate-container scripts (so they're common for all CI jobs) the "old" (built from master prior to this PR) image cannot work with this PR.  @mheon will have to merge this manually.

Once this code is on master, Quay will build a new gate container that will function properly.  However, during the interumn time before it's built, this task will simply fail everywhere.

OTOH, I think having commonly used paths everywhere is a worthwhile change to suffer a few failures over.

Make sense?

**Comment by edsantiago** [comment]
>Make sense?

I guess. I agree with the elimination of conflicting/duplicate/inconsistent `$GOSRC` (although there's still one unfortunate duplication that I don't know how to get rid of). I just have a hard time understanding all the container setup and wasn't sure if this error was "something got missed" or "it will get fixed when the magic container gets rebuilt". I will trust your judgment that it's the latter.  Thanks.

File: contrib/gate/Dockerfile (hunk 3 above)
Comment (edsantiago) [reply contrib/gate/Dockerfile]: My priorities are maintainability >> optimizations, but I won't block because of this.

**Comment by edsantiago** [comment]
LGTM except for the `mkdir` optimization. Nice work.

**Comment by cevich** [comment]
@mheon @baude PTAL, this one needs a manual merge.  I've manually tested the new gate container as best as I can, but there's no simple way to get it the "old" gate image to pass with this specific PR.

**Comment by mheon** [comment]
@cevich Do we have a way to back this out quickly if it goes bad? Want to make sure we aren't going to end up with a red master and no plan to resolve it

**Comment by cevich** [comment]
@mheon The backout plan is: Revert this commit.  Allow quay.io to notice and build the "old" gate container.

Backout plan B: I will make a copy of the current gate container tagged "reverted_5039"

**Comment by mheon** [comment]
@cevich Ack. Let's wait until I cut 1.8.0 and then get this in.

**Comment by cevich** [comment]
yep, np.

Tag created: quay.io/libpod/gate:reverted_5039

**Comment by cevich** [comment]
@mheon we released 1.8 already no?

**Comment by mheon** [comment]
1.8.0 is out, 1.8.1 is in RC

**Comment by mheon** [comment]
So yes, we should probably go ahead with this

**Comment by edsantiago** [comment]
>1.8.1 is in RC

Um, how certain are you? RC1 had rootless regressions, and there haven't been any builds since then...

**Comment by mheon** [comment]
We cut an RC2 tag, didn't we?

**Comment by mheon** [comment]
https://github.com/containers/libpod/releases/tag/v1.8.1-rc2

Must not have been built?

**Comment by edsantiago** [comment]
fc31 build [failed in koji](https://koji.fedoraproject.org/koji/buildinfo?buildID=1472551), I saw no other attempted builds

**Comment by cevich** [comment]
All well and good, but **THIS** pr is needed to improve the gate job accuracy and documentation :smile:   It's also got some important fixes to avoid running container images produced from other branches.

**Comment by edsantiago** [comment]
Way over my head, but LGTM in principle.
